File: backend/app/models/agenda.py
from enum import StrEnum

from pydantic import BaseModel


class AgendaItemType(StrEnum):
    section = "section"
    item = "item"
    speech = "speech"


class AgendaItemIn(BaseModel):
    item_type: AgendaItemType
    title: str
    break_minutes: int | None = None
    duration_green_sec: int | None = None
    duration_yellow_sec: int | None = None
    duration_red_sec: int | None = None
    start_time_override: str | None = None  # "HH:MM"
    host_member_id: str | None = None
    host_name: str | None = None
    evaluator_member_id: str | None = None
    evaluator_name: str | None = None
    path_code: str | None = None
    level_project: str | None = None


class AgendaItemOut(AgendaItemIn):
    id: str
    meeting_id: str
    position: int
    computed_start_time: str  # "HH:MM" — server-computed via the time cascade


class AgendaSaveIn(BaseModel):
    items: list[AgendaItemIn]
    word_of_day: str | None = None
    word_of_day_meaning: str | None = None
    word_of_day_usage: str | None = None
    idiom_of_day: str | None = None
    idiom_of_day_meaning: str | None = None
    idiom_of_day_usage: str | None = None


class AgendaOut(BaseModel):
    items: list[AgendaItemOut]
    word_of_day: str | None = None
    word_of_day_meaning: str | None = None
    word_of_day_usage: str | None = None
    idiom_of_day: str | None = None
    idiom_of_day_meaning: str | None = None
    idiom_of_day_usage: str | None = None

File: web/src/types/agenda.ts
export type AgendaItemType = 'section' | 'item' | 'speech';

export interface AgendaItem {
  id: string;
  meeting_id: string;
  position: number;
  item_type: AgendaItemType;
  title: string;
  break_minutes: number | null;
  duration_green_sec: number | null;
  duration_yellow_sec: number | null;
  duration_red_sec: number | null;
  start_time_override: string | null; // "HH:MM"
  host_member_id: string | null;
  host_name: string | null;
  evaluator_member_id: string | null;
  evaluator_name: string | null;
  path_code: string | null;
  level_project: string | null;
  computed_start_time: string; // "HH:MM"
}

// The editable shape — server-assigned fields are dropped; a new draft row
// has no computed_start_time until the cascade util fills it in locally.
export type AgendaItemDraft = Omit<AgendaItem, 'id' | 'meeting_id' | 'position' | 'computed_start_time'>;

export interface Agenda {
  items: AgendaItem[];
  word_of_day: string | null;
  word_of_day_meaning: string | null;
  word_of_day_usage: string | null;
  idiom_of_day: string | null;
  idiom_of_day_meaning: string | null;
  idiom_of_day_usage: string | null;
}

export interface AgendaSavePayload {
  items: AgendaItemDraft[];
  word_of_day: string | null;
  word_of_day_meaning: string | null;
  word_of_day_usage: string | null;
  idiom_of_day: string | null;
  idiom_of_day_meaning: string | null;
  idiom_of_day_usage: string | null;
}

export function emptyDraft(item_type: AgendaItemType): AgendaItemDraft {
  return {
    item_type,
    title: '',
    break_minutes: item_type === 'section' ? 0 : null,
    duration_green_sec: item_type === 'section' ? null : 0,
    duration_yellow_sec: item_type === 'section' ? null : 0,
    duration_red_sec: item_type === 'section' ? null : 0,
    start_time_override: null,
    host_member_id: null,
    host_name: null,
    evaluator_member_id: null,
    evaluator_name: null,
    path_code: null,
    level_project: null,
  };
}
